### PR TITLE
Expression Evaluator Cleanup 2 - Unify Precomputed Values

### PR DESCRIPTION
**Pull request by csarofeen**
Unify FusionPrecomputedValues and KernelPrecomputedValues
Before:
```
"LayerNormBackward_ShapeInference",355,39.8235,39.8231,us,,,,,
"LayerNormForward_ShapeInference",1124,12.0098,12.0086,us,,,,,
"LayerNormBackward_NoShapeInferenceCachedBaseline",757,18.3266,18.3248,us,,,,,
"LayerNormForward_NoShapeInferenceCachedBaseline",3260,4.35221,4.3517,us,,,,,
```
After:
```
"LayerNormBackward_ShapeInference",360,38.4194,38.4172,us,,,,,
"LayerNormForward_ShapeInference",1095,12.4742,12.4732,us,,,,,
"LayerNormBackward_NoShapeInferenceCachedBaseline",767,18.2251,18.2236,us,,,,,
"LayerNormForward_NoShapeInferenceCachedBaseline",2930,4.4089,4.40844,us,,,,,
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/expr_evaluator.h [hunk 1]
```diff
@@ -26,13 +26,16 @@ class TORCH_CUDA_CU_API ExpressionEvaluator : private OptInConstDispatch {
   //! Bind a concrete value to a named scalar
   void bind(const std::string& name, const IntOrDouble& concrete_value);
 
+  //! Set a concrete value for a parallel dimension
+  void bind(ParallelType pt, Int::ScalarType concrete_value);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/expr_evaluator.h]: This shouldn't have been pulled in here. Haven't added its definition yet. Will remove in this PR, and follow up in the next.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 2]
```diff
@@ -248,7 +248,7 @@ c10::optional<int64_t> Val::getInt() const {
 }
 
 c10::optional<double> Val::getDouble() const {
-  if (isConstScalar() && isAnInt()) {
+  if (isConstScalar() && isADouble()) {
     if (this->getValType() == ValType::Scalar) {
       if (this->isA<Double>()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Should remove these two lines above.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Which lines are you referring to? Lines 251 and 252? Wouldn't we still need to have `isConstScalar()`?

**Review by naoyam — APPROVED** [review]
LGTM.

**Comment by csarofeen** [comment]
> Which lines are you referring to? Lines 251 and 252? Wouldn't we still need to have isConstScalar()?

Yes.

**Review by zasdfgbnm — APPROVED** [review]

**Review by zasdfgbnm — CHANGES_REQUESTED** [review]
Seeing some failure after rebase:
```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedPointwise_CUDA
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedPointwiseMergeContig_CUDA
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedPointwiseMergeSymbolicPass_CUDA
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedRFactor_CUDA
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedStride_CUDA
[  FAILED  ] NVFuserTest.FusionVectorizeMisalignedStrideFail_CUDA
```
Looking into it.

**Comment by csarofeen** [comment]
O.o that's a lot of conflicts today. Strange I didn't see any failures

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**